examples/maths-automarker.py

```diff
@@ -0,0 +1,148 @@
+import json
+from operator import itemgetter
+from typing import TypedDict
+from sympy.parsing.latex import parse_latex
+from sympy import simplify, Integral
+
+import requests
+import typer
+
+
+class MarkPayload(TypedDict):
+    mark: float
+    feedback: str
+
+
+students_url = "/students"
+questions_url = "/questions"
+
+
+def lookup_key(*tokens):
+    return "-".join(map(str, tokens))
+
+
+def build_lookup_table(items):
+    keys = lambda i: [k for k in ["question", "part", "section", "task"] if k in i]
+    return {lookup_key(*itemgetter(*keys(i))(i)): i for i in items}
+
+
+def make_request(url, method="get", params=None, data=None):
+    res = getattr(requests, method)(url, params=params, data=data)
+    if res.status_code == requests.codes.ok:
+        return res.json()
+    raise Exception(f"Request for '{url}' failed with status code {res.status_code}")
+
+
+# ^ Helpers ^ ==========================================================================================
+# You should not need to modify the above functions.
+
+# Example of a question structure
+# instructions: |
+#       Calculus questions
+#     i:
+#       instructions: |
+#         Consider the function $x^2$.
+#       maximum mark: 30
+#       tasks:
+#       - instructions: |
+#           Enter the integral.
+#         type: maths single answer
+#       - instructions: |
+#           Enter the derivative.
+#         type: maths single answer
+
+maths_single_answer_marks = {
+    "1-3-1-1": { "answer": "x^3/3 + C", "mark": 20},
+    "1-3-1-2": { "answer": "2x", "mark": 10},
+}
+
+
+model_answer = r"\frac{\pi}{2} \tanh \frac{\pi}{2}"
+
+class Automarker:
+    def __init__(self, marks: dict[str, dict], answers: dict[str, dict]):
+        self.marks = marks
+        self.answers = answers
+
+    def run(self, username: str, section_id: str, max_mark: int, tasks: list[dict]):
+        if self.marks.get(section_id) is None:
+            mark = 0
+            has_maths_answer = False
+
+            for t, task in enumerate(tasks, 1):
+                if task["type"].startswith("MATHS_SINGLE_ANSWER"):
+                    has_maths_answer = True
+                    task_id = lookup_key(section_id, t)
+                    answer = self.answers.get(task_id, {}).get("answer")
+                    task_mark_and_answer = maths_single_answer_marks[task_id]
+                    answer = json.loads(self.answers.get(task_id, {}).get("answer"))["latex"]
+                    if task_mark_and_answer and answer:
+                        expr1 = parse_latex(answer)
+                        expr2 = parse_latex(task_mark_and_answer["answer"])
+
+                        feedback = "Awarded designated marks for answer."
+                        if (expr1.has(Integral)):
+                            feedback = "Solution should not contain an integral"
```

Comment:
If one of the tasks has Integral in the expression, feedback is set to "Solution should not contain an integral". However, it looks like that gets overwritten on the following loop cycle (i.e. on the next task) with "Awarded designated marks for answer". When do you want to use one feedback? When do you intend to use the other? Could it be simpler, for the demonstrative purpose of this script, to stick to the generic "Awarded designated marks for answer" directly on line 94? Obviously, more sophisticated alternatives are possible. I am just reasoning on what's easier to ship in a clean shape